### PR TITLE
Add get_new_indexers_and_screen function

### PR DESCRIPTION
**Pull request by chaburkland**
Implement `get_new_indexers_and_screen`. For use by `static_frame.IndexHierarchy._extract_iloc`.

### CODE REVIEW

**Comment by chaburkland** [comment]
@brandtbucher 

I'm getting segfaults and I'm not sure why. Would you mind giving this a preliminary review to help me discover where I went wrong?

**Comment by chaburkland** [comment]
> @brandtbucher
> 
> I'm getting segfaults and I'm not sure why. Would you mind giving this a preliminary review to help me discover where I went wrong?

I figured out what my issue was, so there are no more segfaults! However, I would still appreciate a review if you have time :)

**Comment by flexatone** [comment]
@chaburkland : what as the issue that caused the segfaults?

**Comment by chaburkland** [comment]
> @chaburkland : what was the issue that caused the segfaults?

While I'm not 100% sure, I believe it's because my implementation has a **strong** assumption that the incoming array is int64, as we do a C-cast on the underlying data buffer. This check was not validated, and so for certain architectures where the default integers are not int64, I think the cast was leading to instances where invalid memory was being accessed.

I'm not sure exactly where the failure was happening, but it appears all the segfaults happened on 32bit architectures (or Windows), and after I updated all my examples to use int64, and added an explicit type check at the beginning of the function, everything compiled successfully.

**Review by brandtbucher — COMMENTED** [review]
I can't speak to the correctness of the algorithm, but to me this looks mostly good. I've added a couple of random thoughts inline.

More broadly: `num_unique_pyint` and `AK_get_index_screen` are a bit smelly to me. In general, I worry when I see `PyObject_GetItem` and `PyLong_FromLong` calls in what is otherwise supposed to be a get-out-of-Python C accelerator for array stuff. Other than your main loop, this seems mostly like it's just "Python but in C". My intuition when looking at this would be to try to get rid of boxed `PyLongObject`s and `PyObject_GetItem` calls wherever possible, right?

In other words, nothing's stopping us from doing the Python stuff in Python and only using a C extension for the critical lines 756-785, right?

File: src/_arraykit.c [hunk 1]
```diff
@@ -491,6 +491,318 @@ isna_element(PyObject *Py_UNUSED(m), PyObject *arg)
...
+        Py_DECREF(found_mask);
+        return NULL;
+    }
+
+    // 3. Argsort ``found_element_locations``
+    // Quicksort is safe since array just contains integers
+    PyObject* order_found = PyArray_ArgSort(
+            (PyArrayObject*)found_element_locations, // array
+            0,                                       // axis
+            NPY_QUICKSORT                            // kind
```
Comment (flexatone) [inline src/_arraykit.c:555]: I do not see the availability of this argument in the public API docs!
https://numpy.org/doc/stable/reference/c-api/array.html#c.PyArray_ArgSort

File: test/test_util.py [hunk 2]
```diff
@@ -382,6 +382,44 @@ def test_dtype_from_element_str_and_bytes_dtypes(self) -> None:
             self.assertEqual(np.dtype(f'|S{size}'), dtype_from_element(bytes(size)))
             self.assertEqual(np.dtype(f'<U{size}'), dtype_from_element('x' * size))
 
+    def test_get_new_indexers_and_screen_a(self) -> None:
```
Comment (flexatone) [inline test/test_util.py:385]: I suggest more tests that call `get_new_indexers_and_screen()` with mixtures of positional args and kwargs and kwargs only.

File: src/_arraykit.c [hunk 3]
```diff
@@ -491,6 +491,318 @@ isna_element(PyObject *Py_UNUSED(m), PyObject *arg)
...
+        return NULL;
+    }
+
+    // We use ``num_unique`` to signal that we haven't found the element yet
+    // This works, because each element must be 0 < num_unique.
+    if (PyArray_FillWithScalar(element_locations, num_unique_pyint)) {
+        Py_DECREF(element_locations);
+        Py_DECREF(num_unique_pyint);
+        return NULL;
+    }
```
Comment (flexatone) [inline src/_arraykit.c:738]: I think all of this can be done without using `PyLong_FromLong`, and instead building a C array of size equal to `element_locations` * bytes per int64, and then just passing that array as `data` to PyArray_SimpleNewFromData. Something like this:

npy_int64 data[num_unique] = { num_unique };
PyArrayObject *element_locations = PyArray_SimpleNewFromData(1, npy_intp_dims, NPY_INT64, data)

**Review by flexatone — APPROVED** [review]
Nice work on this! I appreciate the thorough documentation and examples.

File: src/_arraykit.c (hunk 1 above)
Comment (chaburkland) [reply src/_arraykit.c:555]: That is very strange! I see the argument defined here, and everything is worked.

https://github.com/numpy/numpy/blob/main/numpy/core/src/multiarray/item_selection.c#L1419

File: src/_arraykit.c [hunk 4]
```diff
@@ -491,6 +491,318 @@ isna_element(PyObject *Py_UNUSED(m), PyObject *arg)
     Py_RETURN_FALSE;
 }
 
+//------------------------------------------------------------------------------
+// get_new_indexers_and_screen & related
+
+static PyObject *
+AK_get_index_screen(PyObject *element_locations,
+                    PyObject *positions,
```
[the remark this answers is not in the document]
Comment (chaburkland) [reply src/_arraykit.c:499]: Per Brandt's feedback, I think I will remove this function entirely, and just write it in Python.

However, I have a different perspective. I think it unhelpful to pass them in as `PyArrayObjects`, as the function never needs to know that. To me it's similar to typehinting a function as `def func(arg: List):` versus `def func(arg: Iterable)`. In this function, we only need these arguments as `PyObject`, so I believe that's how the function should be defined.

If I see a function takes `PyArrayObject` arguments, that communicates to me as the reader that the function _needs to know_ that in order to perform it's operations. In this case, that information is not needed, so I don't want it provided.

File: src/_arraykit.c (hunk 3 above)
Comment (chaburkland) [reply src/_arraykit.c:738]: This array will need to be returned to caller in some cases, so I am concerned about where this data lives. In your example, the data is on the stack, and thus will be de-allocated once the function is exited, leading to segfaults.

If I create the array using `malloc`, then `delete` will never be called because I don't know when the array will be garbage collected.

Comment (chaburkland) [reply src/_arraykit.c:738]: In general, I'm not sure the cost of creating one additional PyObject is worth the extra code complexity

**Comment by chaburkland** [comment]
```python
<class '__main__.GetNewIndexersAndScreenPerf'>
cls                             func             ak              ref              ref/ak     
GetNewIndexersAndScreenPerf     ordered          0.56844831      5.97792536       10.51621616
GetNewIndexersAndScreenPerf     unordered        0.42423423      10.39049229      24.49234776
GetNewIndexersAndScreenPerf     tiled            0.38999344      8.84025321       22.6676972 
GetNewIndexersAndScreenPerf     repeat           0.42471759      7.42528738       17.48288185
GetNewIndexersAndScreenPerf     quick_exit       0.07150431      4.37841167       61.23283912
GetNewIndexersAndScreenPerf     late_exit        0.73353364      11.87805548      16.19292536
GetNewIndexersAndScreenPerf     small            0.53452164      10.94190227      20.470457  
GetNewIndexersAndScreenPerf     large            0.27184065      5.28212683       19.43096713  
```

File: src/_arraykit.c (hunk 3 above)
Comment (flexatone) [reply src/_arraykit.c]: Here is a questionably better way to fill that array that lets NumPy manage the buffer's memory correctly. At least it is a low fewer lines of code!

```c
PyObject *array = PyArray_Zeros(1, dims, dtype, 0)
npy_int64 *array_buffer = (npy_int64*)PyArray_DATA((PyArrayObject*)array);
npy_int64 *end = array_buffer + count;
while (array_buffer < end) {*array_buffer++ = num_unique}
```

File: src/_arraykit.c [hunk 5]
```diff
@@ -491,6 +491,318 @@ isna_element(PyObject *Py_UNUSED(m), PyObject *arg)
...
+    if (PyArray_FillWithScalar(element_locations, num_unique_pyint)) {
+        Py_DECREF(element_locations);
+        Py_DECREF(num_unique_pyint);
+        return NULL;
+    }
+
+
+    PyArrayObject *new_indexers = (PyArrayObject*)PyArray_Empty(
+            1,                                // ndim
+            PyArray_DIMS(array),              // shape
```
[the remark this answers is not in the document]
Comment (chaburkland) [reply src/_arraykit.c]: Yeah this could segfault if a buggy caller does that. I'll add a check.

**Comment by chaburkland** [comment]
```python
<class '__main__.GetNewIndexersAndScreenPerf'>
cls                             func             ak              ref              ref/ak
GetNewIndexersAndScreenPerf     ordered          65.87289547     1156.38919865    17.55485607
GetNewIndexersAndScreenPerf     unordered        41.42570534     2004.67702913    48.39210371
GetNewIndexersAndScreenPerf     tiled            33.83376436     1739.00276589    51.39844173
GetNewIndexersAndScreenPerf     repeat           36.48492943     1423.43519807    39.01433333
GetNewIndexersAndScreenPerf     quick_exit       5.54228165      855.40508353     154.341684
GetNewIndexersAndScreenPerf     late_exit        64.83247354     2305.70365384    35.56402414
GetNewIndexersAndScreenPerf     small            37.22550323     2132.38490796    57.28290347
GetNewIndexersAndScreenPerf     large            33.07465854     1029.98867426    31.1413245
```

**Comment by flexatone** [comment]
@chaburkland : if you think this is ready lets merge it and get a new release out.
